### PR TITLE
feature(rjy): add simple policy head demo

### PR DESCRIPTION
**Pull request by nighood**

### CODE REVIEW

File: obs_encoder.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def __init__(self, obs_space:ObservationSpace, output_size=256, hidden_size=256)
             'supporter': self.supporter_encoder
         }
         # merge
-        obs_input_sizes = {
+        self.obs_merge_input_sizes = {
```
Comment (PaParaZz1) [inline obs_encoder.py:169]: why name `obs_merge_input` rather than `obs_input`

File: policy_head.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+class ActionArgHead(nn.Module):
+    r"""
+    Overview:
+        
+    Interfaces:
+        ``__init__``, ``forward``
+    """
+    def __init__(
+        self,
+        encoded_part_obs_shape,
```
Comment (PaParaZz1) [inline policy_head.py]: add typing lint

File: policy_head.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        self.W_q = nn.Linear(action_type_logit_shape, hidden_size)
+
+    def forward(
+        self,
+        obs_embedding,  # (B, hidden_size)
+        action_type_logit,   # (B, action_type_num)
+        encoded_part_obs     # shape (B, args_shape, encoded_part_obs_shape), part of encoded_obs related to the current arg_head
+        ):
+        # cross attention
+        key = self.W_k(encoded_part_obs)   # (B, args_shape, hidden_size)
```
Comment (PaParaZz1) [inline policy_head.py:36]: norm `key` vector

File: policy_head.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    action_type_names = {getattr(ActionType, attr): attr for attr in dir(ActionType) if not callable(getattr(ActionType, attr)) and not attr.startswith("__")}
+    # action_obs_name_map is used to match action names and encoded_obs names for action_args_head
+    action_obs_name_map = {'play_card':'card_obs','use_skill':'skill_obs', 'change_character': 'character_obs'}
+    def __init__(
+        self,
+        obs_embedding_shape: Union[int, SequenceType],
+        action_shape: Union[int, SequenceType, EasyDict],
+        action_space: ActionSpace,
+        encoded_obs_shape: Dict,    # Should correspond to obs_merge_input_sizes in ObservationEncoder
+        # actor_head_hidden_size: int = 64,
```
Comment (PaParaZz1) [inline policy_head.py]: why comment this

File: policy_head.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    r"""
+    Overview:
+        The VAC model for DI-Card.
+    Interfaces:
+        ``__init__``, ``forward``, ``compute_actor``, ``compute_critic``
+    """
+    mode = ['compute_actor', 'compute_critic', 'compute_actor_critic']
+    # action_type_names is a dict corresponding to the sequence number and action type name
+    # e.g {'play_card':0}
+    action_type_names = {getattr(ActionType, attr): attr for attr in dir(ActionType) if not callable(getattr(ActionType, attr)) and not attr.startswith("__")}
```
Comment (PaParaZz1) [inline policy_head.py]:
```python
from dataclasses import asdict
action_type_names = asdict(ActionType())
```

File: policy_head.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        ) -> Dict:
+        # sample_action_type could be 'argmax' or 'normal'
+        assert sample_action_type in ["argmax", "normal"], "sample_action_type should be 'argmax' or 'normal'"
+        action_type_logit = self.actor_action_type(obs_embedding)['logit']
+
+        action_args_logit = {}
+        if train_mode:
+            # If it is training mode, output action_type and the distribution of a single action_arg
+            # selected by the corresponding sampling method
+            action_type = torch.multinomial(action_type_logit, 1).item() if sample_action_type == 'normal'\
```
Comment (PaParaZz1) [inline policy_head.py]: you need to first transform logit to prob then use multinomial sample

File: policy_head.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                obs_embedding=obs_embedding,
+                action_type_logit=action_type_logit,
+                encoded_part_obs=encoded_obs[select_encoded_obs_name]
+            )
+        else:
+            for action_type_name in self.action_obs_name_map.keys():
+                encoded_obs_name = self.action_obs_name_map[action_type_name]
+                action_args_logit[action_type_name] = self.actor_action_args[action_type_name](
+                    obs_embedding=obs_embedding,
+                    action_type_logit=action_type_logit,
```
Comment (PaParaZz1) [inline policy_head.py]: add a softmax for action_type_logit before passed into action args head

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: obs_encoder.py (hunk 1 above)
Comment (nighood) [reply obs_encoder.py:169]: Because I want to emphasize that this is the size of each encoded_obs before entering the merge fusion, I think it will be used in the policy_head, which is easy to distinguish from the obs_embedding after fusion

File: policy_head.py (hunk 4 above)
Comment (nighood) [reply policy_head.py]: Because arg_head is three (play_card/use_skill/change_character), each arg_head needs the shape information of the intermediate result encoded_obs of the encoder (of course, the modified encoder now) when initializing, namely obs_merge_input_sizes. I will change it later.

File: policy_head.py [hunk 8]
```diff
@@ -115,30 +119,39 @@ def compute_actor(
...
         train_mode = False,
+        selected_action_type =None,     # action_type selected outside
         ) -> Dict:
         # sample_action_type could be 'argmax' or 'normal'
         assert sample_action_type in ["argmax", "normal"], "sample_action_type should be 'argmax' or 'normal'"
         action_type_logit = self.actor_action_type(obs_embedding)['logit']
-
+        action_type_prob = torch.softmax(action_type_logit, dim=-1)
+        if selected_action_type is not None:
+            action_type = selected_action_type  # 这里作用域不知道有没有问题
```
Comment (PaParaZz1) [inline policy_head.py]: 没问题

File: policy_head.py [hunk 9]
```diff
@@ -155,31 +168,41 @@ def compute_actor_critic(
...
+                # 检查所选action_type是否有action_args
+                select_encoded_obs_name = self.action_obs_name_map[select_action_name]
+                action_args_logit[select_action_name] = self.actor_action_args[select_action_name](
+                    obs_embedding=obs_embedding,
+                    action_type_prob=action_type_prob,
+                    encoded_part_obs=encoded_obs[select_encoded_obs_name]
+                )
+            else:
+                # 没有或需要用规则计算action_args。应该返回None还是空字典？
+                action_args_logit = None
```
Comment (PaParaZz1) [inline policy_head.py:192]: 返回None

File: policy_head.py [hunk 10]
```diff
@@ -155,31 +168,41 @@ def compute_actor_critic(
...
         action_type_logit = self.actor_action_type(obs_embedding)['logit']
+        action_type_prob = torch.softmax(action_type_logit, dim=-1)
 
+        if selected_action_type is not None:
+            action_type = selected_action_type  # 这里作用域不知道有没有问题
+        else:
+            action_type = torch.multinomial(action_type_prob, 1).item() if sample_action_type == 'normal'\
+                            else torch.argmax(action_type_prob, 1).item()
         action_args_logit = {}
         if train_mode:
```
Comment (PaParaZz1) [inline policy_head.py]: 这里用 `self.training`, 这是 nn.Module 基类中就维护好的变量，以及 `self.training` 为True则三个args head都计算，为False则只计算action type对应的head

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: obs_encoder.py [hunk 11]
```diff
@@ -239,7 +258,7 @@ def process_obs(self, obs_list:list, last_action_list):
             ], dim=0)   # (10)
             processed_obs_lists['dice_obs'].append(dice_obs)
 
-            character_obs = torch.cat([
+            character_obs = torch.stack([
```
Comment (PaParaZz1) [inline obs_encoder.py:260]: you can directly stack with dim=1 to avoid the following transpose

File: obs_encoder.py [hunk 12]
```diff
@@ -137,16 +137,24 @@ class ObservationEncoder(nn.Module):
     Overview:
         Encoder that processes and encodes the overall obs.
     """
-    def __init__(self, obs_space:ObservationSpace, output_size=256, hidden_size=256):
+    def __init__(
+            self,
+            obs_space:ObservationSpace,
+            output_size=256,
+            # embedding_num=256
```
Comment (PaParaZz1) [inline obs_encoder.py]: remove commented code

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
